File: facilitator-solutions/go/rung-2-sqrt-bound.go
package main

import "math"

// Rung 2 - stop the outer loop at the square root.
//
// Go starts on a flat []bool already, so there is no rung 1 here - the growable
// list rung is a C# and JavaScript problem.
//
// Every composite has a prime factor no larger than its own square root, so once
// the factor passes sqrt(n) every composite below n has already been marked by
// something smaller. What this removes is the outer loop's scan of all 1,000,000
// entries looking for the next prime. The saving is a read pass, not writes.

type primeSieve []bool

func getSieve(sieveSize int) primeSieve {
	sieve := make([]bool, sieveSize+1)
	for i := range sieve {
		sieve[i] = true
	}
	return sieve
}

func (sieve primeSieve) run() {
	sieveSize := len(sieve) - 1
	limit := int(math.Sqrt(float64(sieveSize)))

	for factor := 2; factor <= limit; factor++ {
		if sieve[factor] {
			q := factor + factor
			for q <= sieveSize {
				sieve[q] = false
				q += factor
			}
		}
	}
}

func (sieve primeSieve) result() []int {
	var primes []int

	for i := 2; i < len(sieve); i++ {
		if sieve[i] {
			primes = append(primes, i)
		}
	}

	return primes
}

File: facilitator-solutions/fsharp/rung-2-sqrt-bound.fs
namespace DragRace

// Rung 2 - stop the outer loop at the square root. Includes rung 1.
//
// Every composite has a prime factor no larger than its own square root, so once
// the factor passes sqrt(n) there is nothing left to mark.
//
// What this removes is the outer loop's scan of all 1,000,000 entries looking for
// the next prime. The saving is a read pass, not writes.
//
// Remember this array is 1-based against the numbers: index i holds the number
// i + 1, so Array.length candidates IS n.
module Sieve =
    let create sieveSize = Array.create sieveSize true

    let sieveOut factor candidates =
        let rec markNonPrimes factor q =
            if q > Array.length candidates then
                candidates
            else
                candidates.[q - 1] <- false
                markNonPrimes factor (q + factor)

        markNonPrimes factor (factor + factor)

    let run candidates =
        let limit = int (sqrt (float (Array.length candidates)))

        let rec run' factor =
            if factor > limit then
                candidates
            else
                if candidates.[factor - 1] then
                    sieveOut factor candidates |> ignore

                run' (factor + 1)

        run' 2

    let getResults candidates =
        let rec getResults' candidates current primes =
            if current > Array.length candidates
            then primes
            else if candidates.[current - 1]
            then getResults' candidates (current + 1) (current :: primes)
            else getResults' candidates (current + 1) primes

        getResults' candidates 2 [] |> List.rev
